### DISCUSSION TITLE
read/write Bytestring NodesIds

### DISCUSSION THREAD

**Discussion by dostojewskyi** [FreeOpcUa/opcua-asyncio#1434] [Q&A] [answered]
Hi all, 
I try to read/write Nodes from an OPCua Server which only has ByteString Node IDs.
I'm able to get_nodes('ns=x;<ByteString>'). But always when calling get_value() or get_data_value() I receive errors, that not value exist on the decoded Bytestring Node. Or I'm even noch able to read any value from any node. 
Looking at the OPCua Server Side, I see, that the addressed Nodes have Values to represent.

**Comment by schroeder-** [comment]
Pass a bytestring instead of a string:
`get_nodes(b'ns=x;')`

**Reply by dostojewskyi** [reply]
already tried this. 
```
try:
 n = client.get_node(b"ns=3;b=EBA=") #bytetring
 logger.info(f'node: {n}')
```
raises error Exception can't get node 

```
try: 
 n= client.get_node("ns=3;b=EBA=") #string
...
try: 
 await n.read_data_value()
```
raises error Exception no Value on node ns=3;b=0x4542413d

using Bytestring at get_node() also hints for type error ```Expected type 'Node | NodeId | str | int', got 'bytes' instead``` 

asyncua version is 1.0.4

**Reply by schroeder-** [reply, marked as answer]
Sry my bad. 
`client.get_node(ua.NodeId(b'EBA=', 3))`

**Reply by dostojewskyi** [reply]
I do remember coming from using ua.NodeId. 
But receiving the same issue: 
```[__main__    ] INFO     no value on node ns=3;b=0x4542413d```

this happens with:
```client.get_node(ua.NodeId(b'EBA=', 3, ua.NodeIdType.ByteString))``` and also w/o NodeIdType ```client.get_node(ua.NodeId(b'EBA=', 3))```

I have the feeling I do misunderstand something completely. But I have no clue what I'm currently lacking. 
this is the frontend of my OPC server GUI looking for this node: 
<img width="459" alt="image" src="https://github.com/FreeOpcUa/opcua-asyncio/assets/58233550/4742caa6-eb8b-4b89-80bb-ed775b06ae56">
should be allowing read & histRead
<img width="481" alt="image" src="https://github.com/FreeOpcUa/opcua-asyncio/assets/58233550/7fc3f34d-18cc-4929-91bc-0d79133607da">

**Reply by dostojewskyi** [reply]
so found something. 
I create a new Variable on Server Side w/ Opaque type.. 
I name it: ```0x4542413d``` which ends up in ```RUJBPQ==``` - I can use ```add_node(b'EBA=')``` but not ```add_node(b'RUJBPQ==')```, nor ```add_node(b'0x4542413d')```. But I only have the list of ```RUJBPQ==``` typed elements. How can I encode them back to ```EBA=``` - can I find the logic of this somewhere?

screenshot of the Node I have created. 
<img width="615" alt="image" src="https://github.com/FreeOpcUa/opcua-asyncio/assets/58233550/933c85de-67a6-4d88-9b80-a13b14d26317">

code snippet:
```
n = client.get_node(ua.NodeId(b'EBA=', 3, ua.NodeIdType.ByteString))
V = await n.read_value()
```
logs 
```
INFO     node: ns=3;b=0x4542413d
INFO     value: -0.7516713
```

**Reply by schroeder-** [reply]
You can convert a hex string to a bytes like this:
`bytes.fromhex('4542413d')` -> b'EBA='
Is there a reason why using oqaque instead of other identifiers? Because if you want it readable use string and if you want speed use integer as nodeid.

**Reply by dostojewskyi** [reply]
our historian only hosts opaque as identifier.   
So reverse engineered I have ```RUJBPQ==``` as ID, which I can call as ```EBA=``` which again is represented in a hex string as ```4542413d``` this hex string is same for both. 

Since on Server Side it is shown as ```RUJBPQ==``` and on Client side it's called as ```EBA=```, and I only have the ```RUJBPQ==``` representation, how do I get to the ```EBA=``` representation (which is for all other nodes unknown to me)?

**Reply by dostojewskyi** [reply]
SOLVED: 
OPCUA Server encodes on different base then Client.
So each known NodeId is base64, which needs to be decoded:

```
import base64
base64.b64decode(b'<bytestring>')
```

nevertheless, thank you @schroeder- your input was valuable to better understand some of the logics behind.
